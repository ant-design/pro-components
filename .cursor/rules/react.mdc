---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
