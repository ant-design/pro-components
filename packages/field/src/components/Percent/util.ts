/** 获取展示符号 */
export function getSymbolByRealValue(realValue: number) {
  if (realValue === 0) {
    return null;
  }
  if (realValue > 0) {
    return '+';
  }
  return '-';
}

/** 获取颜色 */
export function getColorByRealValue(realValue: number /** ,color: string */) {
  if (realValue === 0) {
    return '#595959';
  }
  return realValue > 0 ? '#ff4d4f' : '#52c41a';
}

/** 获取到最后展示的数字 */
export function getRealTextWithPrecision(
  realValue: number,
  precision: number = 2,
) {
  return precision >= 0 ? realValue?.toFixed(precision) : realValue;
}

/**
 * 转化为数字
 * @copy from https://github.com/toss/es-toolkit/blob/32a183828c244d675f46810935e45dfefec81a54/src/compat/util/toNumber.ts#L19
 */
export function toNumber(value: any): number {
  if (typeof value === 'symbol' || value instanceof Symbol) {
    return NaN;
  }

  return Number(value);
}
