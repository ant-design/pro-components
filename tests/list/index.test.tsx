import ProList, { BaseProList } from '@ant-design/pro-list';
import {
  cleanup,
  fireEvent,
  render as reactRender,
  screen,
  waitFor,
} from '@testing-library/react';
import { Tag } from 'antd';
import type { Key } from 'react';
import { act, useState } from 'react';
import PaginationDemo from '../../packages/list/src/demos/pagination';
import { waitForWaitTime } from '../util';

type DataSourceType = {
  name: string;
  desc: {
    text: string;
  };
};

afterEach(() => {
  cleanup();
});

describe('List', () => {
  it('🚏 base use', async () => {
    const { container } = reactRender(
      <ProList
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
        }}
      />,
    );
    expect(
      container.querySelector('.ant-pro-list-row-title')!.innerHTML,
    ).toEqual('我是名称');
    expect(
      container.querySelector('.ant-pro-list-row-description')!.innerHTML,
    ).toEqual('desc text');
  });

  it('🚏 BaseList', async () => {
    const { container } = reactRender(
      <BaseProList
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
          xxx: {
            dataIndex: ['desc', 'text'],
          },
          subTitle: {
            title: 'desc text',
          },
        }}
      />,
    );
    expect(
      container.querySelector('.ant-pro-list-row-title')!.innerHTML,
    ).toEqual('我是名称');
    expect(
      container.querySelector('.ant-pro-list-row-description')!.innerHTML,
    ).toEqual('desc text');
    expect(container.querySelectorAll('.ant-pro-card')!.length).toBe(0);
  });

  it('🚏 show loading state', async () => {
    const { container } = reactRender(
      <ProList
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
        }}
        loading={true}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('🚏 only has content', async () => {
    const { container } = reactRender(
      <ProList
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          content: {
            render: () => {
              return (
                <div>
                  段落示意：蚂蚁金服设计平台
                  design.alipay.com，用最小的工作量，无缝接入蚂蚁金服生态，提供跨越设计与开发的体验解决方案。蚂蚁金服设计平台
                  design.alipay.com，用最小的工作量，无缝接入蚂蚁金服生态提供跨越设计与开发的体验解决方案。
                </div>
              );
            },
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('🚏 only has description', async () => {
    const { container } = reactRender(
      <ProList
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          description: {
            render: () => (
              <>
                <Tag>语雀专栏</Tag>
                <Tag>设计语言</Tag>
                <Tag>蚂蚁金服</Tag>
              </>
            ),
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('🚏 empty', async () => {
    const { container } = reactRender(
      <ProList
        metas={{
          title: {
            dataIndex: 'name',
          },
        }}
      />,
    );
    expect(
      container.querySelector('.ant-empty-description')!.innerHTML,
    ).toEqual('暂无数据');
  });

  it('🚏 expandable', async () => {
    const onExpand = vi.fn();
    const Wrapper = () => {
      const [expandedRowKeys, onExpandedRowsChange] = useState<readonly Key[]>(
        [],
      );
      return (
        <ProList
          dataSource={[
            {
              name: '我是名称',
              content: <div>我是内容</div>,
            },
          ]}
          metas={{
            title: {
              dataIndex: 'name',
            },
            content: {},
          }}
          expandable={{ expandedRowKeys, onExpandedRowsChange, onExpand }}
        />
      );
    };
    const { container } = reactRender(<Wrapper />);
    expect(
      container.querySelectorAll('.ant-pro-list-row-description').length,
    ).toEqual(0);
    await fireEvent.click(
      container.querySelector('.ant-pro-list-row-expand-icon')!,
    );
    expect(
      container.querySelector('.ant-pro-list-row-content')!.innerHTML,
    ).toEqual('<div>我是内容</div>');
    expect(onExpand).toHaveBeenCalledWith(
      true,
      expect.objectContaining({ name: '我是名称' }),
    );
  });

  it('🚏 expandable support expandRowByClick', async () => {
    const onExpand = vi.fn();
    const Wrapper = () => {
      const [expandedRowKeys, onExpandedRowsChange] = useState<readonly Key[]>(
        [],
      );
      return (
        <ProList
          dataSource={[
            {
              name: '我是名称',
              content: <div>我是内容</div>,
            },
          ]}
          metas={{
            title: {
              dataIndex: 'name',
            },
            content: {},
          }}
          expandable={{
            expandedRowKeys,
            onExpandedRowsChange,
            onExpand,
            expandRowByClick: true,
          }}
        />
      );
    };
    const { container } = reactRender(<Wrapper />);
    expect(
      container.querySelectorAll('.ant-pro-list-row-description').length,
    ).toEqual(0);
    await fireEvent.click(container.querySelector('.ant-list-item')!);
    expect(
      container.querySelector('.ant-pro-list-row-content')!.innerHTML,
    ).toEqual('<div>我是内容</div>');
    expect(onExpand).toHaveBeenCalledWith(
      true,
      expect.objectContaining({ name: '我是名称' }),
    );
  });

  it('🚏 expandable with defaultExpandedRowKeys', async () => {
    const Wrapper = () => {
      return (
        <ProList
          dataSource={[
            {
              name: '我是名称',
              content: <div>我是内容</div>,
              itemKey: 'a',
            },
            {
              name: '我是名称',
              content: <div>我是内容b</div>,
              itemKey: 'b',
            },
          ]}
          rowKey="itemKey"
          metas={{
            title: {
              dataIndex: 'name',
            },
            content: {},
          }}
          expandable={{
            defaultExpandedRowKeys: ['b'],
          }}
        />
      );
    };
    const { container } = reactRender(<Wrapper />);
    expect(
      container.querySelector('.ant-pro-list-row-content')!.innerHTML,
    ).toEqual('<div>我是内容b</div>');
  });

  it('🚏 expandable with expandedRowRender', async () => {
    const Wrapper = () => {
      const [expandedRowKeys, onExpandedRowsChange] = useState<readonly Key[]>(
        [],
      );
      return (
        <ProList
          dataSource={[
            {
              name: '我是名称',
              content: <div>我是内容</div>,
            },
          ]}
          metas={{
            title: {
              dataIndex: 'name',
            },
            content: {},
          }}
          expandable={{
            expandedRowKeys,
            onExpandedRowsChange,
            expandedRowClassName: () => {
              return 'test-custom-class-name';
            },
            expandedRowRender: (record, index) => {
              return <div>expand:{index}</div>;
            },
          }}
          rowKey={(item) => {
            return item.name;
          }}
        />
      );
    };
    const { container } = reactRender(<Wrapper />);
    expect(
      container.querySelectorAll('.ant-pro-list-row-description').length,
    ).toEqual(0);
    // html.find('.ant-pro-list-row-expand-icon').simulate('click');
    await fireEvent.click(
      container.querySelector('.ant-pro-list-row-expand-icon')!,
    );
    expect(
      container.querySelector(
        '.ant-pro-list-row-content .test-custom-class-name',
      )!.innerHTML,
    ).toEqual('<div>expand:0</div>');
  });

  it('🚏 expandable with expandIcon', async () => {
    const fn = vi.fn();
    const Wrapper = () => {
      return (
        <ProList
          dataSource={[
            {
              name: '我是名称',
              content: <div>我是内容</div>,
            },
          ]}
          metas={{
            title: {
              dataIndex: 'name',
            },
            content: {},
          }}
          expandable={{
            expandIcon: ({ record }) => (
              <div
                id="test_click"
                onClick={() => fn(record.name)}
                className="expand-icon"
              />
            ),
          }}
          rowKey={(item) => {
            return item.name;
          }}
        />
      );
    };
    const { container } = reactRender(<Wrapper />);

    expect(container.querySelectorAll('.expand-icon')).toHaveLength(1);

    await fireEvent.click(container.querySelector('#test_click')!);
    expect(fn).toHaveBeenCalledWith('我是名称');
  });

  it('🚏 ProList support renderItem', async () => {
    const Wrapper = () => {
      return (
        <ProList
          dataSource={[
            {
              name: '我是名称',
              content: <div>我是内容</div>,
            },
          ]}
          renderItem={(_, index) => {
            return <div data-testid="test_index">{index}</div>;
          }}
          rowKey={(item) => {
            return item.name;
          }}
        />
      );
    };
    reactRender(<Wrapper />);

    expect(screen.getByTestId('test_index')).toHaveTextContent('0');
  });

  it('🚏 rowSelection', async () => {
    const Wrapper = () => {
      return (
        <ProList
          dataSource={[
            {
              name: '我是名称',
              description: '我是描述',
            },
            {
              name: '我是名称',
              description: '我是描述',
            },
          ]}
          rowSelection={{}}
          metas={{
            title: {
              dataIndex: 'name',
            },
            description: {},
          }}
        />
      );
    };
    const { container } = reactRender(<Wrapper />);

    expect(container.querySelectorAll('.ant-checkbox-input')!.length).toEqual(
      2,
    );

    fireEvent.change(container.querySelectorAll('.ant-checkbox-input')[0], {
      target: {
        checked: true,
      },
    });

    expect(container.querySelectorAll('.ant-checkbox-input')[0]).toBeChecked();
    expect(
      container.querySelectorAll('.ant-checkbox-input')[1],
    ).not.toBeChecked();
  });

  it('🚏 support pagination', async () => {
    const { container } = reactRender(<PaginationDemo />);

    expect(container.querySelectorAll('.ant-list-item').length).toEqual(5);

    fireEvent.click(container.querySelectorAll('.ant-pagination-item')[1]);

    expect(container.querySelectorAll('.ant-list-item').length).toEqual(2);

    fireEvent.mouseDown(container.querySelector('.ant-select-selector')!);
    fireEvent.click(container.querySelectorAll('.ant-select-item-option')[3]);

    expect(container.querySelectorAll('.ant-list-item').length).toEqual(7);
  });

  it('🚏 filter and request', async () => {
    const onRequest = vi.fn();
    const { container, findByText, baseElement } = reactRender(
      <ProList<any, { title: string }>
        metas={{
          title: {
            title: '标题',
          },
        }}
        request={(params, sort, filter) => {
          if (params.title) {
            onRequest(params, sort, filter);
          }
          return Promise.resolve({
            success: true,
            data: [
              {
                title: '测试标题1',
              },
              {
                title: '测试标题2',
              },
            ],
          });
        }}
        pagination={{
          pageSize: 5,
          onShowSizeChange: () => {},
        }}
        search={{
          filterType: 'light',
        }}
      />,
    );

    await waitFor(async () => {
      expect(
        container.querySelectorAll('.ant-pro-list-row-title').length,
      ).toEqual(2);
    });

    fireEvent.click(container.querySelector('.ant-pro-core-field-label')!);
    act(() => {
      fireEvent.change(baseElement.querySelector('.ant-input')!, {
        target: {
          value: 'test',
        },
      });
    });

    await act(async () => {
      (await findByText('确 认')).click();
    });

    await waitFor(() => {
      expect(onRequest).toHaveBeenCalledWith(
        {
          current: 1,
          pageSize: 5,
          title: 'test',
        },
        {},
        {},
      );
    });
  });

  it('🚏 ProList support onRow', async () => {
    const onClick = vi.fn();
    const onMouseEnter = vi.fn();
    const { container } = reactRender(
      <ProList<DataSourceType>
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
        }}
        onRow={(record: DataSourceType) => {
          return {
            onMouseEnter: () => {
              onMouseEnter(record.name);
            },
            onClick: () => {
              onClick();
            },
          };
        }}
      />,
    );

    fireEvent.click(container.querySelector('.ant-list-item')!);

    expect(onClick).toBeCalled();

    fireEvent.mouseEnter(container.querySelector('.ant-list-item')!);

    expect(onMouseEnter).toHaveBeenCalledWith('我是名称');
  });

  it('🚏 ProList support rowClassName as a string', async () => {
    const customizedRowClassName = 'rowClassName';
    const { container } = reactRender(
      <ProList
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
        }}
        rowClassName={customizedRowClassName}
      />,
    );

    expect(container.querySelector('li.ant-pro-list-row')!).toHaveClass(
      customizedRowClassName,
    );
    expect(container).toMatchSnapshot();
  });

  it('🚏 ProList support rowClassName as a function', async () => {
    const customizedRowClassName = (_: any, index: number): string =>
      index % 2 === 0 ? 'even' : 'odd';
    const { container } = reactRender(
      <ProList
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
        }}
        rowClassName={customizedRowClassName}
      />,
    );

    expect(container.querySelectorAll('li.ant-pro-list-row')[0]).toHaveClass(
      'even',
    );
    expect(container.querySelectorAll('li.ant-pro-list-row')[1]).toHaveClass(
      'odd',
    );
    expect(container).toMatchSnapshot();
  });

  it('🚏 ProList support itemHeaderRender', async () => {
    const html = reactRender(
      <ProList<DataSourceType>
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        itemHeaderRender={(item) => <>qixian:{item.name}</>}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
        }}
      />,
    );

    await waitForWaitTime(1200);
    expect(
      html.baseElement.textContent?.includes('qixian:我是名称'),
    ).toBeTruthy();
  });

  it('🚏 ProList support itemTitleRender', async () => {
    const html = reactRender(
      <ProList<DataSourceType>
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
          },
        ]}
        itemTitleRender={(item) => <>qixian:{item.name}</>}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
        }}
      />,
    );

    await waitForWaitTime(1200);

    expect(
      html.baseElement.textContent?.includes('qixian:我是名称'),
    ).toBeTruthy();
  });

  it('🚏 list support actions render to extra props', async () => {
    const html = reactRender(
      <ProList
        grid={{ gutter: 16, column: 2 }}
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
            actions: [
              <a key="edit" id="html_url">
                修复
              </a>,
            ],
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
          actions: {},
        }}
      />,
    );

    await waitForWaitTime(1200);
    // 触发click，执行一下 stopPropagation 的代码
    await act(async () => {
      (await html.findByText('修复'))?.click();
    });
    expect(html.baseElement.textContent?.includes('修复')).toBeTruthy();
    expect(
      !!html.baseElement.querySelector('.ant-pro-card-actions'),
    ).toBeFalsy();
  });

  it('🚏 list support actions render to actions props', async () => {
    const html = reactRender(
      <ProList
        grid={{ gutter: 16, column: 2 }}
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
            actions: {},
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
          actions: {
            cardActionProps: 'actions',
            render: () => [
              <a key="edit" id="edit">
                修复
              </a>,
            ],
          },
        }}
      />,
    );
    await waitForWaitTime(1000);

    expect(!!html.baseElement.querySelector('.ant-pro-card-extra')).toBeFalsy();

    act(() => {
      html.queryByText('修复')?.click();
    });
  });

  it('🚏 trigger list item event when has grid prop', async () => {
    const fn1 = vi.fn();
    const fn2 = vi.fn();
    const html = reactRender(
      <ProList
        grid={{ gutter: 16, column: 2 }}
        onItem={(record: any) => {
          return {
            onMouseEnter: () => {
              fn1(record.name);
            },
            onClick: () => {
              fn2(record.name);
            },
          };
        }}
        dataSource={[
          {
            name: '我是名称',
            desc: {
              text: 'desc text',
            },
            actions: {},
          },
        ]}
        metas={{
          title: {
            dataIndex: 'name',
          },
          description: {
            dataIndex: ['desc', 'text'],
          },
          actions: {
            cardActionProps: 'actions',
            render: () => [
              <a key="edit" id="edit">
                修复
              </a>,
            ],
          },
        }}
      />,
    );
    await waitForWaitTime(1000);

    act(() => {
      fireEvent.mouseEnter(
        html.baseElement.querySelector(
          '.ant-pro-list-row-card .ant-pro-checkcard',
        )!,
        {},
      );
      fireEvent.click(
        html.baseElement.querySelector(
          '.ant-pro-list-row-card .ant-pro-checkcard',
        )!,
        {},
      );
    });

    await waitFor(() => {
      expect(fn1).toHaveBeenCalledWith('我是名称');
      expect(fn2).toHaveBeenCalledWith('我是名称');
    });
  });
});
