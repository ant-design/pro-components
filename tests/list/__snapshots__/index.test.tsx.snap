// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > 🚏 ProList support rowClassName as a function 1`] = `
<div>
  <div
    class="ant-pro-table ant-pro-list ant-pro-list ant-pro-list-no-split"
  >
    <div
      class="ant-pro-card"
    >
      <div
        class="ant-pro-card-body"
        style="padding: 0px;"
      >
        <div
          class="ant-list ant-list-split ant-pro-list-container"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <ul
                class="ant-list-items"
              >
                <li
                  class="ant-list-item ant-pro-list-row even"
                >
                  <div
                    class="ant-pro-list-row-header"
                  >
                    <div
                      class="ant-pro-list-row-header-option"
                    />
                    <div
                      class="ant-list-item-meta"
                    >
                      <div
                        class="ant-list-item-meta-content"
                      >
                        <h4
                          class="ant-list-item-meta-title"
                        >
                          <div
                            class="ant-pro-list-row-header-container"
                          >
                            <div
                              class="ant-pro-list-row-title"
                            >
                              我是名称
                            </div>
                          </div>
                        </h4>
                        <div
                          class="ant-list-item-meta-description"
                        >
                          <div
                            class="ant-pro-list-row-description"
                          >
                            desc text
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
                <li
                  class="ant-list-item ant-pro-list-row odd"
                >
                  <div
                    class="ant-pro-list-row-header"
                  >
                    <div
                      class="ant-pro-list-row-header-option"
                    />
                    <div
                      class="ant-list-item-meta"
                    >
                      <div
                        class="ant-list-item-meta-content"
                      >
                        <h4
                          class="ant-list-item-meta-title"
                        >
                          <div
                            class="ant-pro-list-row-header-container"
                          >
                            <div
                              class="ant-pro-list-row-title"
                            >
                              我是名称
                            </div>
                          </div>
                        </h4>
                        <div
                          class="ant-list-item-meta-description"
                        >
                          <div
                            class="ant-pro-list-row-description"
                          >
                            desc text
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`List > 🚏 ProList support rowClassName as a string 1`] = `
<div>
  <div
    class="ant-pro-table ant-pro-list ant-pro-list ant-pro-list-no-split"
  >
    <div
      class="ant-pro-card"
    >
      <div
        class="ant-pro-card-body"
        style="padding: 0px;"
      >
        <div
          class="ant-list ant-list-split ant-pro-list-container"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <ul
                class="ant-list-items"
              >
                <li
                  class="ant-list-item ant-pro-list-row rowClassName"
                >
                  <div
                    class="ant-pro-list-row-header"
                  >
                    <div
                      class="ant-pro-list-row-header-option"
                    />
                    <div
                      class="ant-list-item-meta"
                    >
                      <div
                        class="ant-list-item-meta-content"
                      >
                        <h4
                          class="ant-list-item-meta-title"
                        >
                          <div
                            class="ant-pro-list-row-header-container"
                          >
                            <div
                              class="ant-pro-list-row-title"
                            >
                              我是名称
                            </div>
                          </div>
                        </h4>
                        <div
                          class="ant-list-item-meta-description"
                        >
                          <div
                            class="ant-pro-list-row-description"
                          >
                            desc text
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`List > 🚏 only has content 1`] = `
<div>
  <div
    class="ant-pro-table ant-pro-list ant-pro-list ant-pro-list-no-split"
  >
    <div
      class="ant-pro-card"
    >
      <div
        class="ant-pro-card-body"
        style="padding: 0px;"
      >
        <div
          class="ant-list ant-list-split ant-pro-list-container"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <ul
                class="ant-list-items"
              >
                <li
                  class="ant-list-item ant-pro-list-row"
                >
                  <div
                    class="ant-pro-list-row-header"
                  >
                    <div
                      class="ant-pro-list-row-header-option"
                    />
                  </div>
                  <div
                    class="ant-pro-list-row-content"
                  >
                    <div>
                      段落示意：蚂蚁金服设计平台 design.alipay.com，用最小的工作量，无缝接入蚂蚁金服生态，提供跨越设计与开发的体验解决方案。蚂蚁金服设计平台 design.alipay.com，用最小的工作量，无缝接入蚂蚁金服生态提供跨越设计与开发的体验解决方案。
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`List > 🚏 only has description 1`] = `
<div>
  <div
    class="ant-pro-table ant-pro-list ant-pro-list ant-pro-list-no-split"
  >
    <div
      class="ant-pro-card"
    >
      <div
        class="ant-pro-card-body"
        style="padding: 0px;"
      >
        <div
          class="ant-list ant-list-split ant-pro-list-container"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <ul
                class="ant-list-items"
              >
                <li
                  class="ant-list-item ant-pro-list-row"
                >
                  <div
                    class="ant-pro-list-row-header"
                  >
                    <div
                      class="ant-pro-list-row-header-option"
                    />
                    <div
                      class="ant-list-item-meta"
                    >
                      <div
                        class="ant-list-item-meta-content"
                      >
                        <div
                          class="ant-list-item-meta-description"
                        >
                          <div
                            class="ant-pro-list-row-description"
                          >
                            <span
                              class="ant-tag"
                            >
                              语雀专栏
                            </span>
                            <span
                              class="ant-tag"
                            >
                              设计语言
                            </span>
                            <span
                              class="ant-tag"
                            >
                              蚂蚁金服
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`List > 🚏 show loading state 1`] = `
<div>
  <div
    class="ant-pro-table ant-pro-list ant-pro-list ant-pro-list-no-split"
  >
    <div
      class="ant-pro-card"
    >
      <div
        class="ant-pro-card-body"
        style="padding: 0px;"
      >
        <div
          class="ant-list ant-list-split ant-list-loading ant-pro-list-container"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div>
              <div
                aria-busy="true"
                aria-live="polite"
                class="ant-spin ant-spin-spinning"
              >
                <span
                  class="ant-spin-dot-holder"
                >
                  <span
                    class="ant-spin-dot ant-spin-dot-spin"
                  >
                    <i
                      class="ant-spin-dot-item"
                    />
                    <i
                      class="ant-spin-dot-item"
                    />
                    <i
                      class="ant-spin-dot-item"
                    />
                    <i
                      class="ant-spin-dot-item"
                    />
                  </span>
                </span>
              </div>
            </div>
            <div
              class="ant-spin-container ant-spin-blur"
            >
              <ul
                class="ant-list-items"
              >
                <li
                  class="ant-list-item ant-pro-list-row"
                >
                  <div
                    class="ant-pro-list-row-header"
                  >
                    <div
                      class="ant-pro-list-row-header-option"
                    />
                    <div
                      class="ant-list-item-meta"
                    >
                      <div
                        class="ant-list-item-meta-content"
                      >
                        <h4
                          class="ant-list-item-meta-title"
                        >
                          <div
                            class="ant-pro-list-row-header-container"
                          >
                            <div
                              class="ant-pro-list-row-title"
                            >
                              我是名称
                            </div>
                          </div>
                        </h4>
                        <div
                          class="ant-list-item-meta-description"
                        >
                          <div
                            class="ant-pro-list-row-description"
                          >
                            desc text
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
